### PR TITLE
feat: runtime validation

### PR DESCRIPTION
**Pull request by harrysolovay**
Will swap out the import specifier pending paritytech/scale-ts#118.

**EDIT**: swapped it out

### CODE REVIEW

File: deps/scale.ts [hunk 1]
```diff
@@ -1 +1,2 @@
-export * from "https://deno.land/x/scale@v0.9.0/mod.ts";
+// export * from "https://deno.land/x/scale@v0.9.0/mod.ts";
+export * from "../../scale-ts/mod.ts";
```
Comment (tjjfvi) [inline deps/scale.ts:2]: ... why have I never done this

File: frame_metadata/scale_info.ts [hunk 2]
```diff
@@ -20,8 +20,12 @@ export const $tys: $.Codec<Ty[]> = $.createCodec({
...
   },
-  _assert() {
-    // TODO #362
+  _assert(assert) {
+    assert.instanceof(this, Array);
+    const value = assert.value as unknown[];
+    for (let i = 0; i < value.length; i++) {
+      $tyId._assert(assert.key(this, i));
+    }
   },
```
Comment (tjjfvi) [inline frame_metadata/scale_info.ts:29]:
```suggestion
  _assert: $.array($.ty)._assert,
```

File: frame_metadata/scale_info.ts [hunk 3]
```diff
@@ -36,8 +40,8 @@ export const $tyId: $.Codec<Ty> = $.createCodec({
     const id = $compactU32._decode(buffer);
     return ctx.tys?.[id] ?? { id } as any;
   },
-  _assert() {
-    // TODO #362
+  _assert(assert) {
+    $.compact($.u32)._assert(assert);
   },
```
Comment (tjjfvi) [inline frame_metadata/scale_info.ts:45]:
```suggestion
  _assert: $.compactU32._assert,
```

File: frame_metadata/scale_info.ts [hunk 4]
```diff
@@ -20,25 +20,23 @@ export const $tys: $.Codec<Ty[]> = $.createCodec({
     }
     return tys;
   },
-  _assert() {
-    // TODO #362
+  _assert(assert) {
+    return $.array($ty)._assert(assert);
```
Comment (tjjfvi) [inline frame_metadata/scale_info.ts:24]:
```suggestion
    $.array($ty)._assert(assert);
```

Comment (harrysolovay) [reply frame_metadata/scale_info.ts:24]: Causes used-before-defined error

File: frame_metadata/scale_info.ts [hunk 5]
```diff
@@ -20,25 +20,23 @@ export const $tys: $.Codec<Ty[]> = $.createCodec({
...
   },
   _decode(buffer) {
     const ctx = buffer.context.get(TyDecodeCtx);
     const id = $compactU32._decode(buffer);
     return ctx.tys?.[id] ?? { id } as any;
   },
-  _assert() {
-    // TODO #362
-  },
+  _assert: $compactU32._assert,
```
Comment (tjjfvi) [inline frame_metadata/scale_info.ts:39]:
```suggestion
  _assert(assert){
    $compactU32._assert(assert.access(this, "id"));
  },
```

**Review by tjjfvi — APPROVED** [review]

**Pull request merged**